crates/sui-types/src/effects/effects_v2.rs

```diff
@@ -0,0 +1,288 @@
+                        Owner::ObjectOwner(_) | Owner::AddressOwner(_)
+                    ));
+                }
+                (ObjectIn::NotExist, ObjectOut::ObjectWrite(..), true, false) => {
+                    // created Move object.
+                }
+                (ObjectIn::NotExist, ObjectOut::PackageWrite(_), true, false) => {
+                    // created Move package or user Move package upgrade.
+                }
+                (ObjectIn::Exist(_), ObjectOut::NotExist, false, false) => {
+                    // wrapped.
+                }
+                (ObjectIn::Exist(_), ObjectOut::NotExist, false, true) => {
+                    // deleted.
+                }
+                (
+                    ObjectIn::Exist((old_version, old_digest)),
+                    ObjectOut::ObjectWrite(new_digest, _),
+                    false,
+                    false,
+                ) => {
+                    // mutated.
+                    assert!(old_version.value() < self.lamport_version.value());
+                    assert_ne!(old_digest, new_digest);
+                }
+                (
+                    ObjectIn::Exist((old_version, _)),
+                    ObjectOut::PackageWrite((new_version, _)),
+                    false,
+                    false,
+                ) => {
+                    // system package upgrade.
+                    assert_eq!(old_version.value() + 1, new_version.value());
+                }
+                _ => {
+                    panic!("Impossible object change: {:?}, {:?}", id, change);
+                }
+            }
+        }
+        // Make sure that gas object exists in changed_objects.
+        let (_, owner) = self.gas_object();
+        assert!(matches!(owner, Owner::AddressOwner(_)));
+
+        for (id, _) in &self.unchanged_shared_objects {
+            assert!(unique_ids.insert(*id));
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+struct ObjectChange {
+    // input_state and output_state are the core fields that's required by
+    // the protocol as it tells how an object changes on-chain.
+    /// State of the object in the store prior to this transaction.
+    input_state: ObjectIn,
+    /// State of the object in the store after this transaction.
+    output_state: ObjectOut,
+
+    // id_created and id_deleted are informational fields useful to tell
+    // more comprehensive story of what happened to some objects.
+    /// Whether this object ID is created in this transaction.
+    id_created: bool,
+    /// Whether this object ID is deleted in this transaction.
+    id_deleted: bool,
+}
+
+type VersionDigest = (SequenceNumber, ObjectDigest);
+
+/// If an object exists (at root-level) in the store prior to this transaction,
+/// it should be Exist, otherwise it's NonExist, e.g. wrapped objects should be
+/// NonExist.
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+enum ObjectIn {
+    NotExist,
```

Comment:
We were having a discussion about whether we wanted to record that a (shared) input was deleted between signing and execution -- does that warrant a `Deleted(SequenceNumber)` variant here?

Reply:
No.
That case is covered in unchanged_shared_objects.
We don't care about signing. Effects is always comparing between before execution vs after execution.

Reply:
If I'm trying to figure out why my transaction failed (and it failed because one of the shared transaction inputs was deleted) would I be able to detect that by looking at the unchanged shared inputs? (Are effects even the right thing to look at for that? If not, where do we direct people?)

Reply:
Yes it would be in unchanged_shared_objects with UnchangedSharedKind::Deleted